Adds a new omega error handler

philipwjones opened the pull request:
This includes

  - an Error class and set of macros for handling errors
  - user and developer documentation
  - an Error unit test
  - a new external module cpptrace for generating stack traces

It is built on top of the Log functionality and relies on that for most of the error messages. The stack trace capability is only available when debug info is present (-g flag or a debug build). This only includes the module and unit test - the capability will be propagated throughout Omega in separate PRs to follow. Tests pass on Chrysalis. Other tests will follow.

Checklist
* [x] Documentation:
  * [x] Design document has been generated and added to the docs
  * [x] User's Guide has been updated
  * [x] Developer's Guide has been updated
  * [x] Documentation has been [built locally](https://e3sm-project.github.io/Omega/omega/develop/devGuide/BuildDocs.html) and changes look as expected
* [x] Testing
  * [x] A comment in the PR documents testing used to verify the changes including any tests that are added/modified/impacted.
  * [x] CTest unit tests for new features have been added per the approved design. 
  * [x] Unit tests have passed. Please provide a relevant CDash build entry for verification.

grnydawn commented on components/omega/test/infra/ErrorTest.cpp:
```diff
@@ -0,0 +1,157 @@
+   // The above calls have reset the error so create a new failure
+   TotalError += Error(ErrorCode::Fail, __LINE__, __FILE__, "Error in driver");
+
+   // Test critical error
+   passError1(TotalError);
+
+   // Finalize environments
+   MPI_Finalize();
+
+   return RetVal;
```
"RetVal is not set by the each tests in this unit-test, so the test always returns the initial value, or passError1(TotalError) exits the test early, preventing any change to the returned RetVal."

philipwjones replied:
Thanks, I forgot to add these for the non-fatal tests - but I also realized after seeing the other FAIL below that this test code is expected to fail because of the critical error abort at the end, so I need to find an alternative way of flagging the other test failures.

philipwjones replied:
Fixed with latest commit

grnydawn commented on components/omega/test/infra/ErrorTest.cpp:
```diff
@@ -0,0 +1,157 @@
+       "   [error] [ErrorTest.cpp:121] Initialized to fail String 1\n";
+   Error TotalError(ErrorCode::Fail, __LINE__, __FILE__,
+                    "Initialized to fail {} {}", "String", 1);
+   if (TotalError.isFail() and TotalError.Msg == ExpectMsg) {
+      LOG_INFO("Error constructor with message: PASS");
+   } else {
+      LOG_ERROR("Error constructor with message: FAIL");
+      LOG_ERROR(" Error constructor with message expected: {}", ExpectMsg);
+      LOG_ERROR(" Error constructor with message actual:   {}", TotalError.Msg);
+   }
```
The following log file messages indicate that this test has actually failed:
```
[* error] [ErrorTest.cpp:124] Error constructor with message: FAIL
[* error] [ErrorTest.cpp:125]  Error constructor with message expected:    [error] [ErrorTest.cpp:121] Initialized to fail String 1

[* error] [ErrorTest.cpp:126]  Error constructor with message actual:      [error] [ErrorTest.cpp:119] Initialized to fail String 1
```

philipwjones replied:
I fixed the proximate cause, but as above, I need to find a better way to trap these.

philipwjones replied:
Fixed with latest commit

grnydawn commented on components/omega/doc/devGuide/Error.md:
````diff
@@ -0,0 +1,139 @@
+To check these returned error codes or any other accumulated errors in a code,
+the following macros can be used:
+```c++
+ERROR_CHECK(ErrorToBeChecked, NewMessage, MsgArgs...);
+ERROR_CHECK_WARN(ErrorToBeChecked, NewMessage, MsgArgs...);
+ERROR_CHECK_CRITICAL(ErrorToBeChecked, NewMessage, MsgArgs...);
````
To me, it was a little unexpected that this macro resets the error, as I didn't get that impression from the macro name itself.

It seems that all error-handling macros start their names with `ERROR_`. I thought it might be clearer to start macro names with a verb. For example, `ERROR_RETURN` could be renamed to `MERGE_ERROR`, and `ERROR_CHECK` to `RESET_ERROR`.

Of course, this is just my personal preference—I'm sharing it for your consideration.

philipwjones replied:
The main purpose of these is to check the error code and print the errors so the current name is consistent with that purpose. The reset is a needed side effect to prevent the same error messages getting printed in a later call.

grnydawn reviewed (changes requested):
Please see my reviews in other sections.

Some names in the error handling design document don't seem to match the implementation, such as `ERROR_SUM` and `defineCode`.

philipwjones commented:
Regarding the design doc - this was a case where the implementation I ended up with deviated from the original design. No plans survive first contact...  The functionality is still there. The `ERROR_SUM` is replaced by the addition and accumulate operator overloads (though maybe it's confusing to mix macros with class operators?). In my original design, I had envisioned a repository of pre-defined error messages, but with an Error class, it's just as easy to use a constructor and store a static instance when that functionality is needed. But I think with the current interfaces, I'm not sure it's necessary to pre-define errors and it was a little problematic if you wanted to add any current information into the error message.

philipwjones commented:
And on your other naming comment - I could certainly change the names to `CHECK_ERROR` and `RETURN_ERROR` for those macros. Would `CRITICAL_ERROR` or `ABORT_ERROR` be better for the critical error macro? I had ERROR first since it was kinda like a class name prefix `Error::`.

grnydawn commented:
> Regarding the design doc - this was a case where the implementation I ended up with deviated from the original design. No plans survive first contact... The functionality is still there. The `ERROR_SUM` is replaced by the addition and accumulate operator overloads (though maybe it's confusing to mix macros with class operators?). In my original design, I had envisioned a repository of pre-defined error messages, but with an Error class, it's just as easy to use a constructor and store a static instance when that functionality is needed. But I think with the current interfaces, I'm not sure it's necessary to pre-define errors and it was a little problematic if you wanted to add any current information into the error message.

I think class-based error handling is more flexible than using statically pre-defined errors, as it allows Omega developers to create new errors more easily. For documentation purposes, we might still need to categorize all Omega errors in some way, so that the error handling system can be more useful in the future.

grnydawn commented:
> And on your other naming comment - I could certainly change the names to `CHECK_ERROR` and `RETURN_ERROR` for those macros. Would `CRITICAL_ERROR` or `ABORT_ERROR` be better for the critical error macro? I had ERROR first since it was kinda like a class name prefix `Error::`.

Having a verb at the beginning is my personal preference, but I’ll defer this to the other Omega developers. Based on that preference, `CRITICAL_ERROR` or `ABORT_ERROR` could be renamed to `CREATE_CRITICAL_ERROR` or `CREATE_ABORT_ERROR`, but that seems too verbose and long without much added benefit. In that sense, either `CRITICAL_ERROR` / `ABORT_ERROR` or `ERROR_CRITICAL` / `ERROR_ABORT` works for me.

grnydawn reviewed (approved):
Approved. The unit test correctly prints error messages in the Omega log file, and the return value is properly considered in the test's return code. The stack trace is also correctly generated in DEBUG mode. The test was done on PM-GPU. I left the renaming of the macro to other reviewers.

brian-oneill commented:
I think `externals/cpptrace` should be added here in the devGuide Quick Start, since this is the list of minimum submodules necessary to build Omega: https://github.com/E3SM-Project/Omega/blob/d22df4f24c00e49a4e445ab14ced6cb0f9fb0e24/components/omega/doc/devGuide/QuickStart.md?plain=1#L80-L85

brian-oneill commented:
I don't have a strong preference on the naming conventions. If you want my vote I'llI guess I lean toward Youngsung about the macros starting with verbs, and I think `ABORT_ERROR` is probably the most obvious about what the macro is doing while remaining concise.

brian-oneill commented:
Ran the ctests successfully on Frontier CPU & GPU as well as Chrysalis. Confimred on Chrysalis the backtrace is successfully generated in Debug mode.

philipwjones commented:
Thanks @grnydawn and @brian-oneill   I went ahead and changed the macro names to `ABORT_ERROR`, `RETURN_ERROR`, `CHECK_ERROR_XXX`.  I also changed the assert and require to `OMEGA_ASSERT` and `OMEGA_REQUIRE` (kinda wanted just ASSERT and REQUIRE, but decided I needed to add the prefix for namespace safety).  I updated the docs appropriately (and added cpptrace to dependent libs in quick start) and I re-tested everything to make sure I didn't break stuff. If I don't hear any complaints, I'll go ahead and merge this later.
